File: app/src/main/java/uk/cybertecpro/infosecfeed/FeedFilter.kt
package uk.cybertecpro.infosecfeed

/** Offline feed filtering shared by the search field and category chips. */
object FeedFilter {
    fun apply(items: List<FeedItem>, category: String?, query: String): List<FeedItem> {
        val categoryItems = category?.let { selected ->
            items.filter { it.category == selected }
        } ?: items
        val needle = query.trim()
        if (needle.isBlank()) return categoryItems
        return categoryItems.filter { item ->
            item.title.contains(needle, ignoreCase = true) ||
                item.summary.contains(needle, ignoreCase = true) ||
                item.source.contains(needle, ignoreCase = true)
        }
    }
}

File: windows/InfoSecFeed.Windows/Services/FeedFilter.cs
using InfoSecFeed.Windows.Models;

namespace InfoSecFeed.Windows.Services;

public static class FeedFilter
{
    public static IReadOnlyList<FeedItem> Apply(
        IEnumerable<FeedItem> items,
        string category,
        string query)
    {
        var normalizedQuery = query.Trim();
        return items.Where(item =>
                (category == Categories.All || item.Category == category) &&
                (normalizedQuery.Length == 0 ||
                 item.Title.Contains(normalizedQuery, StringComparison.OrdinalIgnoreCase) ||
                 item.Summary.Contains(normalizedQuery, StringComparison.OrdinalIgnoreCase) ||
                 item.Source.Contains(normalizedQuery, StringComparison.OrdinalIgnoreCase) ||
                 (item.Severity?.Contains(normalizedQuery, StringComparison.OrdinalIgnoreCase) ?? false)))
            .OrderByDescending(item => item.Rank)
            .ToList();
    }
}
